server/jetstream_api.go

```diff
@@ -1180,13 +1193,31 @@ func (s *Server) jsStreamRestoreRequest(sub *subscription, c *client, subject, r
 				Client: caudit,
 			})
 
-			s.Noticef("Completed %s restore for stream %q in account %q in %v", FriendlyBytes(int64(received)), stream, c.acc.Name, end.Sub(start))
+			s.Noticef("Completed %s restore for stream %q in account %q in %v",
+				FriendlyBytes(int64(received)), stream, c.acc.Name, end.Sub(start))
+
+			// If there is a reply subject on the last EOF, send back the stream info or error status.
+			if reply != _EMPTY_ {
+				var resp = JSApiStreamCreateResponse{ApiResponse: ApiResponse{Type: JSApiStreamCreateResponseType}}
+				if err != nil {
+					resp.Error = jsError(err)
+				} else {
+					resp.StreamInfo = &StreamInfo{Created: mset.Created(), State: mset.State(), Config: mset.Config()}
+				}
+				s.sendInternalAccountMsg(acc, reply, s.jsonResponse(&resp))
+			}
 
 			return
 		}
-		// Append chunk to temp file.
-		tfile.Write(msg)
+		// Append chunk to temp file. Mark as issue if we encounter an error.
+		if n, err := tfile.Write(msg); n != len(msg) || err != nil {
+			tfileError = err
+		}
 		received += len(msg)
+
+		if reply != _EMPTY_ {
+			s.sendInternalAccountMsg(acc, reply, nil)
```

Comment:
On the general subject of logging - since we seem to inform ourselves here on googles practises, we dont need to guess or go by tribal knowledge, its extensively documented in the SRE books 

```
White-box monitoring

    Monitoring based on metrics exposed by the internals of the system, including logs, interfaces
    like the Java Virtual Machine Profiling Interface, or an HTTP handler that emits internal 
    statistics.
```
...

```

Black-Box Versus White-Box

We combine heavy use of white-box monitoring with modest but critical uses of black-box monitoring. 
The simplest way to think about black-box monitoring versus white-box monitoring is that black-box 
monitoring is symptom-oriented and represents active—not predicted—problems: "The system isn’t 
working correctly, right now." White-box monitoring depends on the ability to inspect the innards of 
the system, such as logs or HTTP endpoints, with instrumentation. White-box monitoring therefore 
allows detection of imminent problems, failures masked by retries, and so forth.

Note that in a multilayered system, one person’s symptom is another person’s cause. For example, 
suppose that a database’s performance is slow. Slow database reads are a symptom for the database 
SRE who detects them. However, for the frontend SRE observing a slow website, the same slow 
database reads are a cause. Therefore, white-box monitoring is sometimes symptom-oriented, and 
sometimes cause-oriented, depending on just how informative your white-box is.

When collecting telemetry for debugging, white-box monitoring is essential. If web servers seem 
slow on database-heavy requests, you need to know both how fast the web server perceives the 
database to be, and how fast the database believes itself to be. Otherwise, you can’t distinguish 
an actually slow database server from a network problem between your web server and your 
database.

For paging, black-box monitoring has the key benefit of forcing discipline to only nag a human when a problem is both already ongoing and contributing to real symptoms. On the other hand, for not-yet-occurring but imminent problems, black-box monitoring is fairly useless.
```
https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/

In otherwords to OPERATE THE SERVICE they rely very heavily on white box monitoring - logging being one aspect of that - but for alerting and customer feedback they use black box.

This is exactly what I've been proposing all along, we do not operate anymore a fire and forget best efforts middleware, this is now a database.  Without logging we are building a data one simply cannot operate. Modern operation teams use logging EXTENSIVELY.  This isn't some request to pander to people who refuse to move forward, this is how systems are managed. Even by google.

```
Logging

No one needs to look at this information, but it is recorded for diagnostic or for‐ ensic purposes. The expectation is that no one reads logs unless something else prompts them to do so.
```

early in the original SRE book. SRE book also covers how one should not need to run the system at DEBUG level to get basic key operational insights since enabling DEBUG level will most likely cause more problems than it solves.  The only way to operate NATS is to run it at DEBUG level. Or worse to embed it into another process and do something like this which meticulously elevate these incorrect log levels to appropriate levels- which I have had to do in the past:

```go
// Debugf logs at debug level
func (l Logger) Debugf(format string, v ...interface{}) {
	if format == "Error flushing: %v" || format == "TLS handshake error: %v" {
		l.Errorf(format, v)
		return
	}

	l.log.Debugf(format, v...)
}
```

But even this only works if the logs are there to begin with.


Google makes many tools to solve these problems and offers them as a service https://cloud.google.com/blog/products/devops-sre/keeping-your-functions-functioning-with-monitoring-logging-and-error-reporting (now https://cloud.google.com/logging)

Another area we have to look into is structured logs - we've had tickets about this too - this is key to making logs ingested into log aggregation systems a super valuable resource.

We have to log things this just isn't optional or old style operations, its essential.  I'm happy to help improve logging but you first need to agree that this is something we can do before I do another big PR that doesnt get reviewed in this area :)

Reply:
I will take a another look at what we can do here.